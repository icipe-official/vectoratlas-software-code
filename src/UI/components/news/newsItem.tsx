import React, { useState } from 'react';
import { News } from '../../state/state.types';
import {
  Button,
  CardContent,
  CardMedia,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  Grid,
  IconButton,
  Paper,
  Typography,
} from '@mui/material';
import ReactMarkdown from 'react-markdown';
import EditIcon from '@mui/icons-material/Edit';
import { useRouter } from 'next/router';
import { useDispatch } from 'react-redux';
import { AppDispatch } from '../../state/store';
import { deleteNews } from '../../state/news/actions/news.action';

export const NewsItem = ({
  item,
  isEditor,
  hideMoreDetailsButton,
}: {
  item: News;
  isEditor: boolean;
  hideMoreDetailsButton?: boolean;
}) => {
  const router = useRouter();
  const dispatch = useDispatch<AppDispatch>();

  const handleEditClick = () => {
    router.push('/news/edit?id=' + item.id);
  };

  const handleMoreDetailsClick = () => {
    router.push('/news/article?id=' + item.id);
  };

  const [openDialog, setOpenDialog] = useState(false); // State to control dialog visibility
  const [selectedNewsId, setSelectedNewsId] = useState<string | null>(null); // State for selected species ID

  const handleDelete = (newsId: string) => {
    setSelectedNewsId(newsId); // Set selected species ID
    setOpenDialog(true); // Open the confirmation dialog
  };

  const handleConfirmDelete = () => {
    if (selectedNewsId) {
      dispatch(deleteNews(selectedNewsId)); // Dispatch the delete action
    }
    setOpenDialog(false); // Close the dialog after confirming delete
  };

  const handleCloseDialog = () => {
    setOpenDialog(false); // Close the dialog without deleting
  };

  return (
    <Paper sx={{ margin: 0, marginLeft: '2px' }}>
      <Grid
        container
        spacing={0}
        sx={{ justifyContent: 'center' }}
        className="BannerGrid"
      >
        <Grid item xs={12} md={9} key="content">
          <CardContent className="Content">
            <ReactMarkdown
              components={{
                a: ({ node, ...props }) => (
                  <a style={{ color: 'blue' }} {...props} />
                ),
                h2: ({ node, ...props }) => (
                  <h2 style={{ margin: 0 }} {...props} />
                ),
              }}
            >
              {'## ' + item.title}
            </ReactMarkdown>

            <ReactMarkdown
              components={{
                a: ({ node, ...props }) => (
                  <a style={{ color: 'blue' }} {...props} />
                ),
                p: ({ node, ...props }) => (
                  <p
                    style={{
                      marginTop: 15,
                      marginBottom: 0,
                      textAlign: 'justify',
                    }}
                    {...props}
                  />
                ),
              }}
            >
              {item.summary}
            </ReactMarkdown>
          </CardContent>
          <Grid sx={{ justifyContent: 'end', width: '100%', display: 'flex' }}>
            {!hideMoreDetailsButton && (
              <Button
                variant="outlined"
                onClick={handleMoreDetailsClick}
                className="ViewButton"
              >
                More details
              </Button>
            )}
            {isEditor && (
              <Button
                variant="contained"
                onClick={handleEditClick}
                className="EditButton"
              >
                Edit item
              </Button>
            )}
            {isEditor && (
              <Button
                sx={{ backgroundColor: 'red' }}
                variant="contained"
                onClick={() => handleDelete(item.id as string)}
                className="EditButton"
              >
                Delete item
              </Button>
            )}
          </Grid>
        </Grid>
        <Grid
          item
          xs={12}
          md={3}
          key={item.title}
          sx={{ height: '50vh', maxHeight: '40vh' }}
        >
          <CardMedia
            className="Media"
            image={item.image}
            sx={{ height: '100%', overflow: 'hidden' }}
          ></CardMedia>
        </Grid>
        {/* Dialog Box */}
        <Dialog
          open={openDialog}
          onClose={handleCloseDialog}
          aria-labelledby="alert-dialog-title"
          aria-describedby="alert-dialog-description"
        >
          <DialogTitle id="alert-dialog-title">{'Confirm Delete'}</DialogTitle>
          <DialogContent>
            <DialogContentText id="alert-dialog-description">
              Are you sure you want to delete this News? This action cannot be
              undone.
            </DialogContentText>
          </DialogContent>
          <DialogActions>
            <Button onClick={handleCloseDialog} color="primary">
              Cancel
            </Button>
            <Button onClick={handleConfirmDelete} color="secondary" autoFocus>
              Delete
            </Button>
          </DialogActions>
        </Dialog>
      </Grid>
    </Paper>
  );
};
