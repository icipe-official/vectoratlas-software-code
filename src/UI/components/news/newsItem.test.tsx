import React from 'react';
import { fireEvent, render } from '../../test_config/render';
import { NewsItem } from './newsItem';
import { useRouter } from 'next/router';

jest.mock('next/router', () => ({
  useRouter: jest.fn().mockReturnValue({
    query: {},
    push: jest.fn(),
  }),
}));

jest.mock(
  'react-markdown',
  () =>
    function MockReactMarkdown(props: any) {
      return <div>React markdown mock {JSON.stringify(props)}</div>;
    }
);

jest.mock(
  '@mui/icons-material/Edit',
  () =>
    function EditIconMock() {
      return <div>Edit icon mock</div>;
    }
);

describe('NewsItem component', () => {
  let newsItem;

  beforeEach(() => {
    jest.clearAllMocks();

    newsItem = {
      id: '123-456',
      title: 'test news item',
      summary: 'test summary',
      image: 'image-base64-abc-123',
    };

    (useRouter as jest.Mock).mockReturnValue({
      query: {},
      push: jest.fn(),
    });
  });

  it('renders news item correctly', () => {
    const { wrapper } = render(<NewsItem item={newsItem} />);

    expect(wrapper.container).toMatchSnapshot();
  });

  it('hides more details button if flag is set', () => {
    const { wrapper } = render(
      <NewsItem item={newsItem} hideMoreDetailsButton />
    );

    expect(wrapper.queryByText('More details...')).not.toBeInTheDocument();
  });

  it('directs to article page when more details clicked', () => {
    const { wrapper } = render(<NewsItem item={newsItem} />);
    fireEvent.click(wrapper.getByText('More details...'));

    expect(useRouter().push).toHaveBeenCalledWith('/news/article?id=123-456');
  });

  it('shows editor button if isEditor flag is set', () => {
    const { wrapper } = render(<NewsItem item={newsItem} isEditor />);
    expect(wrapper.getByText('Edit icon mock')).toBeInTheDocument();
  });

  it('does not show editor button if isEditor flag is not set', () => {
    const { wrapper } = render(<NewsItem item={newsItem} />);
    expect(wrapper.queryByText('Edit icon mock')).not.toBeInTheDocument();
  });

  it('directs to edit page when edit button clicked', () => {
    const { wrapper } = render(<NewsItem item={newsItem} isEditor />);
    fireEvent.click(wrapper.getByText('Edit icon mock'));

    expect(useRouter().push).toHaveBeenCalledWith('/news/edit?id=123-456');
  });
});
