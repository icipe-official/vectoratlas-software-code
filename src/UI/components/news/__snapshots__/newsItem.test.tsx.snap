// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsItem component renders news item correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1va6po4-MuiPaper-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container BannerGrid css-1lym95h-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-9 css-b1im03-MuiGrid-root"
      >
        <div
          class="MuiCardContent-root Content css-46bh2p-MuiCardContent-root"
        >
          <div>
            React markdown mock 
            {"components":{},"children":"## test news item"}
          </div>
          <div>
            React markdown mock 
            {"components":{},"children":"test summary"}
          </div>
        </div>
        <div
          class="MuiGrid-root css-r4vg1n-MuiGrid-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium ViewButton css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            More details
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3 css-158gt5v-MuiGrid-root"
      >
        <div
          class="MuiCardMedia-root Media css-mme1y-MuiCardMedia-root"
          role="img"
          style="background-image: url(image-base64-abc-123);"
        />
      </div>
    </div>
  </div>
</div>
`;
